Implement most of chrono::high_resolution_clock

ckormanyos opened the pull request:
Fixes #9 via impl of `chrono::high_resolution_clock`

salkinium commented:
Do you want to add a similiar `system_clock` as `chrono::milliseconds` while you're at it? Then targets without a microsecond timer can also reuse this.
Just not sure which one should be the `steady_clock` then… perhaps the default could be a defaulted macro?

ckormanyos commented:
> Do you want to add a similiar system_clock as chrono::milliseconds while you're at it? Then targets without a microsecond timer can also reuse this.

Yes. That is a good idea. I need a few days to implement and test both, maybe throw in a couple of examples.

> Just not sure which one should be the steady_clock then… perhaps the default could be a defaulted macro?

I like the idea of macro(s) for the clocks. I'll draft all these things out in this PR.

rleh commented:
Now that #12 is merged it would be good if you rebase your branch onto to the current master branch. Then the examples and the new code will get compiled by the CI.

chris-durand commented:
> Do you want to add a similiar `system_clock` as `chrono::milliseconds` while you're at it? Then targets without a microsecond timer can also reuse this.
> Just not sure which one should be the `steady_clock` then… perhaps the default could be a defaulted macro?

The standard mandates that "Objects of type `system_­clock` represent wall clock time". Generally they can't be the same since `steady_clock`, you might have guessed it, has to be steady.

ckormanyos commented:
> would be good if you rebase your branch onto to the current master branch

Thx @rleh. Done.

ckormanyos commented:
>> Do you want to add a similiar system_clock as chrono::milliseconds while you're at it? Then targets without a microsecond timer can also reuse this.

> Yes. That is a good idea. I need a few days to implement and test both, maybe throw in a couple of examples.

I am sort of in the middle of this.

As an initial step, I think it is a good first progress to:

1. Add one clock (`high_resolution_clock`) to `<chrono>`.
2. Make 1 other clock configurable via macro to be aliased to `high_resolution_clock`.
3. Make the timebase of `high_resolution_clock` configurable via macro.
4. Add an example of using `<chrono>`.

Thoughts ... ?

The recent push of mine addresses 1,2,4. Point 3 is still open.

Let's see how CI plays out...

Cc: @chris-durand and @salkinium

ckormanyos commented:
Hi @rleh I can't really see where to add [examples/chrono/chrono.cpp](https://github.com/ckormanyos/avr-libstdcpp/blob/chrono_high_resulution_clock/examples/chrono/chrono.cpp) to the CI Makefile

Any tips...?

chris-durand commented:
> Hi @rleh I can't really see where to add [examples/chrono/chrono.cpp](https://github.com/ckormanyos/avr-libstdcpp/blob/chrono_high_resulution_clock/examples/chrono/chrono.cpp) to the CI Makefile
> 
> Any tips...?

You can add your own Makefile to the chrono subfolder and [the CI job will run it](https://github.com/modm-io/avr-libstdcpp/blob/9ff2f79d5c3dc1c1dd7a38ce319d74109edfc3de/.github/workflows/compile_examples.yml#L21).

ckormanyos commented:
Note: I added CI runs to push in additin to pull-request.

Anyway, this thing should be ready for further discussion/review. CI and chrono-example-build are running green in [my fork's CI](https://github.com/ckormanyos/avr-libstdcpp/actions/runs/1712991186).

ckormanyos commented:
> add your own Makefile to the examples/chrono subfolder

Done. Thx @rleh

rleh commented on .github/workflows/compile_examples.yml:
```diff
@@ -1,6 +1,10 @@
 name: Compile Examples
 
-on: [pull_request]
+on:
+  push:
+    branches:
+      - '**'
+  pull_request:
```
Is there a reason to have the CI running on every push?

I think it is not necessary to use the CI for this, it is more sustainable to execute `make` locally...

salkinium replied:
But it could be interesting to build it on the main branch every week, so that updates to the container avr-gcc get tested against this automatically. Alternatively replace `:latest` with a tag, but this repo is not going to be maintained as often as modm.

ckormanyos replied:
For the moment (but I'm surely flexible to change this later) I have retained run CI on branch pushes.

rleh commented on examples/chrono/Makefile:
```diff
@@ -0,0 +1,48 @@
+NAME=chrono-test
+MCU=atmega328p
+
+# cd /mnt/c/MyGitRepos/avr-libstdcpp/examples/chrono
```
```suggestion
```

rleh commented on examples/chrono/Makefile:
```diff
@@ -0,0 +1,48 @@
+LIB_DIR=../..
+
+INCLUDES= -I$(LIB_DIR)/include
+#INCLUDES=-I/mnt/c/MyGitRepos/real-time-cpp/ref_app/src/util/STL
```
```suggestion
```

rleh commented on examples/chrono/chrono.cpp:
```diff
@@ -0,0 +1,219 @@
+  *reinterpret_cast<volatile port_type*>(mcal::reg::portb) = static_cast<port_type>(*reinterpret_cast<volatile port_type*>(mcal::reg::portb) & static_cast<port_type>(~mcal::reg::bval5));
+
+  // Set the port pin direction to output.
+  *reinterpret_cast<volatile port_type*>(mcal::reg::ddrb) = static_cast<port_type>(*reinterpret_cast<volatile port_type*>(mcal::reg::portb) | mcal::reg::bval5);
```
```suggestion
  *reinterpret_cast<volatile port_type*>(mcal::reg::ddrb) = static_cast<port_type>(*reinterpret_cast<volatile port_type*>(mcal::reg::ddrb) | mcal::reg::bval5);
```

but I don't see the benefit (for a _chrono_ example) over

```suggestion
  DDRB |= (1<<5u);
```

This would have allowed the error to be seen much more quickly.

ckormanyos replied:
> have allowed the error to be seen much more quickly

I was not aware of an error at this time. Did you detect one?

ckormanyos replied:
Handled this with use of _regular_ AVR register-access tyles and idioms via `<avr/io.h>`.

rleh replied:
> I was not aware of an error at this time. Did you detect one?

Your code is equivalent to `DDRB = PORTB | (1<<5)`, which does not make sense. You probably meant `DDRB = DDRB | (1<<5)`.

ckormanyos replied:
> Your code is equivalent to `DDRB = PORTB | (1<<5)`, which does not make sense. You probably meant 

Ugghhh... Sigh... I do that kind of thing all the time.

Well, I believe the entire line was (as below) which _may_ have been correct --- albeit a long line... This style is a comparatively safe, reusable tech for a `volatile` read-modify-write direct-memory-access sequence --- but it's non-atomic, maybe a bit slow.

```C
  // Set the port pin direction to output.
  *reinterpret_cast<volatile port_type*>(mcal::reg::ddrb) = static_cast<port_type>(*reinterpret_cast<volatile port_type*>(mcal::reg::portb) | mcal::reg::bval5);
```

Thanks. Good catch. Could you please also check the revised edition?

These things do get hard to read. And, moving forward, those AVR macros do a neat out-of-sight-hiding of the real direct-memroy-access going on, so I'm cool with either way...

rleh reviewed (commented):
I believe this _chrono_ example becomes more useful if we replace the `mcal::gpt::*` and `reinterpret_cast<>` stuff with the (well known) avr-libc syntax (`DDRB |= (1<<5u)`/...).

ckormanyos commented:
> this chrono example becomes more useful if we replace the mcal::gpt::* and reinterpret_cast<> stuff with the (well known)

Oh sure. I'll do that. Got to remember the name of that header at the moment...

Thanks for reviewing.

ckormanyos commented:
> would have allowed the error to be seen much more quickly.

I was not aware of an error at this time. Did you detect one?

chris-durand commented on examples/chrono/chrono.cpp:
```diff
@@ -0,0 +1,219 @@
+
+} // namespace local
+
+auto main(void) -> int
```
Why is `void` used here in the function parameter list?
As far as I know `int main(void)` and `int main()` would be completely equivalent in C++ (contrary to C).

ckormanyos replied:
> Why is `void` used here in the function parameter list?

Good catch. The form of `main` in the example-under-review is I believe, in fact, wrong.

I will correct it to be one of... (see below).

In C++, the allowed forms are either

```C
int main();
```

```C
int main(int, char*);
```

ckormanyos replied:
oops... Typo in the second form, I meant to write

```C
int main(int, char**);
```

ckormanyos commented:
> Is there a reason to have the CI running on every push?

Good question.

How active is your project? How many cycles do you want to burn?

It depends on the activity and quality-depth of your project. i work in some projects in which even the tinyest defect can lead to big problems. So we run tests on syntax, unit tests, etc. --- the whole CI on each and every push, branch push, PR and sometimes even scheduled builds.

After a bit of thought, I might recommend:

1. Let's say there is an active phase in  avr-libstdcpp for a while, add a few examples and docs. Then run CI on push and PR.
2. If this slows doen a bit, then remove CI on pushes if you like.
3. I see no need for nightlies or similar on this repo since it seems slow.changing.

On the other hand regarding points 1 and 2, to run CI on pushes/or-not. Since CI on this project runs in about a minute, why not.

I would also recommend setting up a test matrix that tests a few C++ standards, maybe start with C++17 and C++20.

It remains to be seen if a a significant number of tests might be added. i would, however, have a bunch of simple ideas if you'd like to pursue these.

chris-durand commented on examples/chrono/chrono.cpp:
```diff
@@ -0,0 +1,219 @@
+  // Read the MCU status register.
+  volatile const std::uint8_t mcu_status_register = *reinterpret_cast<volatile std::uint8_t*>(mcal::reg::mcusr);
+
+  static_cast<void>(mcu_status_register);
```
Why is the status register read here? The result is not used. Does the read have any hardware side effects?

ckormanyos replied:
> Why is the status register read here? The result is not used. Does the read have any hardware side effects?

I forgot and I will check. Looking at my records, I first wrote a version of that code around 2005 in assembly. I'll see if it is needed in newer manuals.

ckormanyos replied:
>> Why is the status register read here? 

> I forgot and I will check. Looking at my records, I first wrote a version of that code around 2005 in assembly. I'll see if it is needed in newer manuals.

Ah... Evidently readint the status register was a placeholder for myself in case I had ever wanted to query the reset reason.

Looking at Sect. 15.8.2 in Microchip's "Atmel-2545W-ATmega48/V / 88/V / 168/V_Datasheet_Complete-11/2016", there seems to be no need to read the MCU status register.

I think, however, it's a good idea to clear the WDRF flag and will do so.

ckormanyos replied:
And finally for the elephant in the room --- why use a watchdog timer at all in a simple example meant to exemplify aspects of using `<chrono>` ?

The possible choices are: Include watchdog? yes/no.

This is optional. When using `<chrono>` an example makes sense it there is a little microcosm of a timer structure running. Whenever timers are involved (and using them for delays), it makes sense to trigger a watchdog.

So I opted for use watchdog: Yes.

ckormanyos commented:
OK folks,

Many, many thanks for the excellent, detailed and very thoughtful review comments.

I believe the recent checkins address all review comments. if anything else is found in further review rounds, please don't hesitate to post mention of it.

I added a test matrix to CI which tests on `-std=c++17` _and_ `-std=c++20`. I also added the `-Werror` flag to the chrono example's CI, as there are some neat new warnings in C++20 available in GCC.

As a final comment, I retained the running of CI on pushes, as I really think this is best if/as/when we may add a few more tests/examples to this repo. I can, however, be convinced otherwise, but running a minute or so on CI, i would like that added security as a contributor to this/these repos.

Cc: @chris-durand and @rleh and @salkinium

salkinium reviewed (commented):
I feel like this is too convoluted an example, it focusses on things like namespacing to hint at a larger HAL library that isn't really there or `auto main() -> int` that is just forced because C++ can do it, but really what value does it add?

I would focus on the chrono now function, and take out the WDT, remove the static_casts in favor of C-style access, move `int main()` to the bottom with a visual //==== separator to find it easier.

This example should be very simple, to convince someone to integrate this into their C code without scaring them away with some unfortunate C++ bloat? What do you think?

ckormanyos commented on examples/chrono/chrono.cpp:
```diff
@@ -0,0 +1,190 @@
+  asm volatile("sei");
+
+  // Set the port pin value to low.
+  PORTB = static_cast<std::uint8_t>(PORTB & static_cast<std::uint8_t>(~(1U << PORTB5)));
```
> This isn't really a hill I'm enthused to die on, but this verbosity isn't helping the reputation of C++ either 

Sigh... that's an annoying one... I'm not overly concerned with it either. C++20 on GCC11 will issue a warning regarding non-conformance with `volatile` read-modify-write sequence in the _standard_ way.

But if you want it that way, we can program it any way you like --- which I'll go ahead and do then shortly after the second change request...

I suppose the result of `~(1U << PORTB5)` is not identical with the register width either.

So to resolve this conversation, please, guys, tell me which you prefer, the loose-and-terse style in the community, or type-safe-volatile-conforming? Then I'll happily program it any way you like. it's your project...

ckormanyos replied:
I suppose `<avr/io.h>` has a little terse pp definition for `asm volatile("sei");` as well. if we go all in for `<avr/io.h>` defs, then I'll replace that as well...

ckormanyos replied:
So directly... Would you like me to go all in on `<avr/io.h>` pp defs? Yes or no? I'm totaly happy either way... Your project, your choice, your style. i'm happy to adhere.

Cc: @salkinium and @rleh

salkinium replied:
Yes, I'd like to use the `<avr/io.h>` and `<avr/interrupt.h>` definitions as much as possible.

C++20 went wrong with deprecating volatile read-modify-write sequences and there's a proposal in C++23 that fixes this for embedded use-cases. Please add the `-Wno-volatile` flag.

chris-durand replied:
It will be un-deprecated again for bitwise operations. `|=`, `&=` and `^=` should be fine in C++23 again.

chris-durand replied:
> I suppose the result of ~(1U << PORTB5) is not identical with the register width either.

That is an annoying one ... `int` promotions inherited from C are what they are. I am all in to mark dangerous narrowing operations with an explicit cast. But here it seems apparent for a human that a value bit-or-ed with a `uint8_t` or `(1u << 5)` fits into the `uint8_t` although the expression types are plain `int`. If every register access is marked with lots of these casts they just lose their effect to point out exceptional, unsafe code. In my opinion it just gets too noisy.

ckormanyos commented on examples/chrono/chrono.cpp:
```diff
@@ -0,0 +1,190 @@
+      | static_cast<std::uint8_t>(1U << WDP1)
+      | static_cast<std::uint8_t>(1U << WDP2)
+      | static_cast<std::uint8_t>(1U << WDE)
+    );
```
The delays in the loop are 1 second. I threw in a watchdog for the sake of _politeness_. We can agree to take it out, as it detracts from and adds no value to the example. I'll kick out the WDT.

Do you want me to disable it in the fuse bits (with no code), or keep it active in fuses and deactivate it subsequently in code? You can do either. Default fuses is WDT deactivated. Please indicate which you prefer.

salkinium replied:
I'm for keeping the default fuses that keep it disabled.

ckormanyos commented:
> This example should be very simple, to convince someone to integrate this into their C code without scaring them away with some unfortunate C++ bloat? What do you think?

@salkinium and guys. Decide which you want. The classic example is a 3-to-5 liner that measures the time of a code sequence.

It might be better to go for that when showcasing only the functionality of  `<chrono>`.

This is a larger refactor/complete-re-design. Please decide. I'm happy to go either way. This post would basically render those above moot. Might be the easiest way to proceed... We could go that way and wrap up `<chrono>` lickety split if that's the decision...?

HJust provide me your feedback. I'm neutral on it...

Cc: @salkinium and @rleh and @chris-durand

salkinium commented:
> The classic example is a 3-to-5 liner that measures the time of a code sequence.

Let's go with that. Simpler and easier is better.

ckormanyos commented:
>> The classic example is a 3-to-5 liner that measures the time of a code sequence.

> Let's go with that. Simpler and easier is better.

You bet. Coming right up.

Compiling it on CI. Just for the sake of completion, here are a couple of those [C++20 `volatile` warnings](https://github.com/ckormanyos/avr-libstdcpp/runs/4926804646?check_suite_focus=true#step:5:8). I'll suppress these and move forward.

ckormanyos commented:
OK CI green.

Pls review.... I'm happy to tune at your request(s)...

I stubbed the clock [here](https://github.com/ckormanyos/avr-libstdcpp/blob/b7c816a2d756ac1ad7d23ad37190e4a207e53ec9/examples/chrono/chrono.cpp#L44).

Cc: @salkinium and @rleh and @chris-durand

ckormanyos commented on examples/chrono/chrono.cpp:
```diff
@@ -0,0 +1,190 @@
+
+  for(;;)
+  {
+    // Toggle the LED via direct memory access.
```
pls. review simplified example. We can even strip away the port toggle even further if desired.

ckormanyos commented on examples/chrono/chrono.cpp:
```diff
@@ -0,0 +1,190 @@
+  for(;;)
+  {
+    // Toggle the LED via direct memory access.
+    PORTB = static_cast<std::uint8_t>(PORTB ^ static_cast<std::uint8_t>(1U << PORTB5));
```
See simplified example below.

salkinium replied:
^= is better for understanding, the PIN trick is not well known.

salkinium commented:
Ok, that's perhaps a little too minimal, I would still keep the TIMER0_OVF code, that was pretty cool with guarding the rollover via a double read.

ckormanyos commented:
> Ok, that's perhaps a little too minimal, I would still keep the TIMER0_OVF code, that was pretty cool with guarding the rollover via a double read.

Ok. That was one of my open questions in my mind also...
I'll draft that out tomorrow and run it on a board... To-be-continued...

salkinium commented on examples/chrono/chrono.cpp:
```diff
@@ -0,0 +1,58 @@
+
+  for(volatile unsigned long toggles = 0UL; toggles < 10000UL; ++toggles)
+  {
+    // Toggle the LED via direct memory access.
```
Direct Memory Access (DMA) is a completely different thing than Memory Mapped Input/Output (MMIO) register access.

ckormanyos replied:
I can change the comment.

salkinium commented on examples/chrono/chrono.cpp:
```diff
@@ -0,0 +1,58 @@
+auto main() -> int
+{
+  // Set the port pin value to low.
+  PORTB |= ~(1U << PORTB5);
```
typo?
```suggestion
  PORTB |= (1U << PORTB5);
```
or 
```suggestion
  PORTB &= ~(1U << PORTB5);
```

ckormanyos replied:
I will finish the draft tomorrow. Thanks. I will check it on a board and fix these typos.

ckormanyos commented:
OK folks the next iteration of the `<chrono>`-example's refactoring is available.

This iteration of refactoring has/does:
- Very simple, readable form. Clarity hopefully improved now.
- Restore the actual use of system tick at 1MHz.
- Incorporate most review points to use `avr/whatever.h`-stuff.
- Include watchdog (but only deactivates it, as delivery state of Arduino has WDT active, I think).
- Still open is toggle mechanism (`PINB` or logical xor?).
- Also if WDT deactivate ramins, then I need help with pneumonic for [this line](https://github.com/ckormanyos/avr-libstdcpp/blob/12a64812fcb00e7867a11c344d34aa290770a200/examples/chrono/chrono.cpp#L54).

Pls. take a look and comment as needed/felt.

Cc: @salkinium and @rleh and @chris-durand

rleh commented on examples/chrono/chrono.cpp:
```diff
@@ -0,0 +1,150 @@
+
+static void app_hw_init();
+
+auto main() -> int
```
I still don't understand the point of this spelling in such a place.

ckormanyos replied:
>  point of this spelling in such a place.

What spelling do you mean? Do you mean `int main()`?

chris-durand replied:
I guess the comment refers to the trailing return type. I am fine with either notation.

ckormanyos replied:
> the trailing return type.

Well it's kind of a funny story. When I started using `clang-tidy`, the most prominent genre of messages I was getting were in the category `modernize-*`. This particular style of modern syntax (the trailing return type) was the message I did, in fact, most prominently receive.

So I went ahead and modernized everything in my style that I learned from those experiences --- thinking that would be best --- although I must admit, we're pretty used to seeing `int main()`

OK, long story, so when I get around to switching over to tabs as line indentations (see also #13), I'll go ahead and use the traditional form for `main()`.

ckormanyos commented on examples/chrono/chrono.cpp:
```diff
@@ -0,0 +1,150 @@
+static void app_hw_init();
+
+auto main() -> int
+{
+  // Initialize the application hardware. This includes WDT, PORTB.5 and TIMER0.
+  app_hw_init();
+
+  for(;;)
+  {
+    // Toggle the LED on portb.5.
```
> ... uses tabs

Oh, here we have our first actual conflict of style. I do not usually use tabs.

But..., it's your project, your style. I'll switch to tabs if that's the style here. No problem.

> add a `.clang-format` file and enforce it using CI?

That would be nice. I also recently started to use `clang-tidy` and I am _very_ happy with the results. I have starte in my own projects to add tidy checks to CI as well.

I think, however, that `clang-format` and `clang-tidy` enforce/check/CI should be undertaken in qa second step, as i'd like to get the `<chrono>` example in there and finally get started/to-work/finished with #10.

After that, I'd actually recommend to start with `clang-tidy`, then move on to format issues.

In the intermediate time, I'll go ahead and switch over to tabs. 

Are tabs the consensus around here?

Cc: @salkinium and @chris-durand

chris-durand replied:
> Oh, here we have our first actual conflict of style. I do not usually use tabs.
>
> But..., it's your project, your style. I'll switch to tabs if that's the style here. No problem.

Tabs is what we use in the modm project because we inherited that from its predecessor xpcc.

chris-durand replied:
:+1: for automatic formatting and static analysis

salkinium replied:
I would prefer spaces over tabs, but I don't want to have a confusing mix of policies on modm, so we're going with tabs for now. clang-format makes it easy to work with anyways.

ckormanyos replied:
If I am here, I will comply. Thx for clarification.

ckormanyos commented on .github/workflows/compile_examples.yml:
```diff
@@ -1,13 +1,22 @@
 
-on: [pull_request]
+on:
+  push:
+    branches:
+      - '**'
```
Awesome! I did not know that. I actually am a real noob in GHA and have a whole bunch to learn... Thx @rleh

chris-durand reviewed (approved):
Nice! Thank you!

ckormanyos commented:
Hi guys, this one is, for the moment, done and good to go, in my opinion.

Tabs as line indentations to be handled in #13. Or should I handle tabs as line indentations here in this PR?

ckormanyos commented:
OK I decided to:

- Handle line indentation as tabs here in this PR.
- Use `wdt_reset()` which is found to be defined as `asm volatile("wdr")` in `<avr/wdt.h>`.

Sorry about the late changes.

salkinium commented:
Squash?

ckormanyos commented:
> Squash?

Hi Niklas (@salkinium), if this comment is for me, I'm not exactly sure how to do it...

Could you please explain more details?

salkinium commented:
The commits in this PR should be squashed into one commit, so that the git history is clean and understandable.

ckormanyos commented:
> The commits in this PR should be squashed into one commit

Well Niklas (@salkinium) I think I did this, or something with essentially the same result. I do admit, though, that I'm not experienced with Git squashing methods.

jwakely commented:
Please consider contributing this clock impl back upstream.

ckormanyos commented:
> Please consider contributing this clock impl back upstream

Hi Jonathon (@jwakely) this is a good idea. We would need to negotiate on the resolution of the clock(s), as I believe this needs to be specified when laying out a concrete implementation.

As time allows, we can/should discuss this matter. I can help where needed and if wanted.
